tools/tool-env/profile-require.ts

```diff
@@ -0,0 +1,185 @@
+// Use path instead of files.ts here because we are explicitly trying to
+// track requires, and files.ts is often a culprit of slow requires.
+import path from 'path';
+
+// seconds since epoch
+const getNow = () => Date.now() / 1000;
+
+let currentInvocation: RequireInvocation;
+
+class RequireInvocation {
+  timeStarted: number;
+  timeFinished: number | null;
+  parent: RequireInvocation;
+  children: RequireInvocation[];
+  selfTime: number | null;
+  totalTime: number | null;
+
+  /**
+   * @param name module required
+   * @param filename file doing the requiring, if known
+   */
+  constructor(
+    public name: string,
+    private filename: string | null = null
+  ) {
+    this.timeStarted = getNow();
+    this.timeFinished = null;
+    this.parent = currentInvocation;
+    this.children = [];
+  
+    this.selfTime = null;
+    this.totalTime = null;
+  }
+
+  isOurCode(): boolean {
+    if (!this.filename) {
+      return this.name === 'TOP';
+    }
+  
+    if (!this.name.match(/\//)) {
+      // we always require our stuff via a path
+      return false;
+    }
+  
+    const ourSource = path.resolve(__dirname);
+    const required = path.resolve(path.dirname(this.filename), this.name);
+    if (ourSource.length > required.length) {
+      return false;
+    }
+    return required.substr(0, ourSource.length) === ourSource;
+  }
+
+  why(): string {
+    let walk: RequireInvocation = this;
+    let last: RequireInvocation | null = null;
+  
+    while (walk && ! walk.isOurCode()) {
+      last = walk;
+      walk = walk.parent;
+    }
+  
+    if (!walk) {
+      return "???";
+    }
+    if (last) {
+      return `${path.basename(walk.name)}:${path.basename(last.name)}`;
+    }
+    return path.basename(walk.name);
+  }
+}
+
+export function start() {
+  const moduleModule = require('module');
+  currentInvocation = new RequireInvocation('TOP');
+
+  const realLoader = moduleModule._load;
+  moduleModule._load = function (...args: [string, { filename?: string }]) {
+    const [id, { filename }] = args;
+    const inv = new RequireInvocation(id, filename);
+    const parent = currentInvocation;
+
+    currentInvocation.children.push(inv);
+    currentInvocation = inv;
+
+    try {
+      return realLoader.apply(this, args);
+    } finally {
+      inv.timeFinished = getNow();
+      currentInvocation = parent;
+    }
+  };
+}
+
+/**
+ * The discrepancy in computed times that we are willing to tolerate
+ * before deciding that the times don't add up.
+ */
+const TOLERANCE = 1/1000;
+
+export function printReport() {
+  currentInvocation.timeFinished = getNow();
+  computeTimes(currentInvocation);
+
+  const summary = summarize(currentInvocation, 0);
+  const times = Object.values(summary).sort((a, b) => 
+    b.time - a.time
+  );
+
+  let ourTotal = 0, otherTotal = 0;
+
+  times.forEach(item => {
+    let line = `${formatTime(item.time)} ${item.name}`;
+
+    if (!item.ours) {
```

Comment:
Sometimes there is a blank line above `if` blocks and other times not. The style guide only says that there should be blank lines after blocks and before new statements: https://github.com/meteor/javascript/#18.6

Since you are converting existing code, I'll just leave this one comment rather than leave 100 comments about where this is inconsistent. But it might be a good opportunity to correct those inconsistencies now:

- Missing blank lines after blocks
- Extra blank lines after statements

Reply:
Hi @jamesmillerburgess ! Thank you for reading through my pull request and for your thoughtful comments.

I agree with the impulse to clean up code (ideally to meet lint and coding style guidelines) as we convert to TypeScript. That said, I honestly don't think we should require ourselves to make existing non-compliant code 100% style guide compliant if it's something that the Meteor team already wasn't prioritizing and if it wasn't already being enforced by a tool, because it means we spend less type actually converting to TypeScript, and more time worrying about clean-up that could hopefully be automated with ESLint/fix and Prettier.

I think most of us are contributing TypeScript conversion PRs as "volunteer" work while also holding full-time jobs, which means our time is a limited resource. And if we can spend more of that time adding statically enforced types and less worrying about cleaning up the style of pre-existing code that already didn't adhere to Meteor's own coding style guide, I think that's a good tradeoff.

As with my other comments, I am obviously willing to be overridden by any member of the Meteor team on this! This is just my subjective opinion, and I grant that it may be controversial. But, adding types is something I'm willing to spend my free time doing because I enjoy it and because I think it's something that's helpful for the Meteor project, which I hope to see succeed (because it adds documentation and type guarantees to the codebase, making it more accessible and hopefully less bug-prone). Adding and removing newlines to old code... to be honest, that's probably not something I'm quite as willing to spend my free time on!

Anyway, my goal isn't to be dismissive here -- I appreciate your comments and pointing out the style guide, and I know that we both have positive intentions. I just want to give some feedback on why I think it definitely would not be productive to leave 100 comments about inconsistent newline placement in existing code (thank you for not doing that).

---

With respect to the specific comment about newlines, I do have this comment to make, though: The fact that the style guide says "leave a blank line after blocks and before the next statement" does _not_ mean "never use newlines in any other circumstance."

Newlines should be used when it's helpful in creating semantic blocks of code, which sometimes means in between statements.

Point taken about missing blank lines after blocks; I will go ahead and update that in this file.

Reply:
Thanks for not taking it personally. I'm not saying I'm right, just leaving points to consider. And I didn't see any real issues with the conversion of this file, so good job!

I wanted to review some of these TypsScript PRs, because they had been here for a while without comment, and I hope we can build some momentum around this conversion.

And thinking about how this TypeScript conversion was presented in @benjamn's Meteor Night talk, the spirit here of this effort is not to clean up every last thing, but to make the code accessible to more developers by clarifying some of the object types that get thrown around a lot.

Reply:
At the end of the day, I know that we both want to see this effort succeed and have Meteor keep on rockin', which is why after I finished writing these long-winded comments, I went ahead and made the changes anyway. 😄 Just wanted to add my POV on why I think it may be okay to not do full cleanup on every file that we convert.

I totally agree about building momentum -- I've been holding off on doing more conversions while waiting for feedback on the ones that are outstanding.